More qt win build fixes

catalinr opened the pull request:
A few other things:
- Probably not all changes are the right ones, but with them I was able to build `wxQT` static libs on Windows and run `widgets` sample. Hopefully more knowledgeable people will review and improve them.
- Dynamic lib build still failed the last time I attempted it.
- Starting with version `5.7`, `Qt` requires `c++11` support, which I enabled explicitly but it would be nice to have it automatically enabled based on the `Qt` version.
- The configure command that I used:
`../configure --with-qt --enable-debug QT5_CUSTOM_DIR=c:/Qt/5.11.0/mingw53_32 --with-cxx=11 --disable-shared`

vadz commented on build/aclocal/bakefile.m4:
```diff
@@ -76,7 +76,7 @@ AC_DEFUN([AC_BAKEFILE_PLATFORM],
     if test "x$BAKEFILE_FORCE_PLATFORM" = "x"; then
         case "${BAKEFILE_HOST}" in
-            *-*-mingw32* )
+            *-*-mingw* )
```
Is this intentional? If so, why?

catalinr replied:
Yes, it extends the accepted values to both `*-*-mingw32*` and `*-*-mingw64*`. The `gcc` version that came with `Qt` identified itself as the latter.
Alternatively the following should work just as well: `*-*-mingw32* | *-*-mingw64* )`.

vadz replied:
I was just wondering how did it always work for us with MinGW64 before, without this change.

What exactly do you mean by "identified"? Is this part of the triplet returned by `config.guess` or something else?

Anyhow, I'm not really against this change, it makes sense. But I just wish I understood why hadn't it been necessary before. And, as usual, it would be better if it could be done in a separate commit.

TIA!

catalinr replied:
No idea about MinGW64.
In my config log there are:
```
configure:3699: checking build system type
configure:3713: result: x86_64-pc-mingw64
configure:3733: checking host system type
configure:3746: result: x86_64-pc-mingw64
```
I don't know if that is given by the compiler or the bash environment but one of those is what seems to be used. Anyway, that change was just a lucky guess.
Later in the same log there is
`gcc.exe (i686-posix-dwarf-rev0, Built by MinGW-W64 project) 5.3.0`.

vadz replied:
The output in `config.log` is from `config.guess`, yes...

vadz commented on configure.in:
```diff
@@ -52,7 +52,7 @@ WX_MSW_VERSION=$wx_major_version_number$wx_minor_version_number$wx_release_numbe
 dnl Autoconf-2.60 changed the meaning of top_builddir variable, so we have
 dnl to use our own replacement that will work with both 2.5x and 2.60+:
-wx_top_builddir="`pwd`"
+wx_top_builddir="`pwd -W 2> /dev/null || pwd`"
```
Hmm, what does `-W` pwd option mean?

catalinr replied:
`pwd -W` produced windows style path using forward slashes, i.e. `c:/my/dir`. It was only available in the Windows version (well, in git bash that is). For Linux there was no `-W` option and the command failed, so the output of plain `pwd` was used.
This path format here was required by `ld`. The same went for the custom dir passed to configure, which upset `ar`. But if there's a better way for obtaining it, no objections here.

vadz replied:
Hmm, interesting, Cygwin bash doesn't have this option. Also pretty weird that MinGW ld and ar don't understand forward slashes. Again, if we have to do this, let's do it, but ideally I'd like to have this in its own commit too, e.g. to be able to revert it easily later if we find a better solution.

catalinr replied:
Well, they did understand fwd slashes, except for the one before the drive letter. They just wouldn't accept paths starting with `/c/` but only `c:/` (mind the colon too).

Yes I'll separate things in different commits in one of the following days.

vadz replied:
Weird, all MinGW stuff is supposed to understand `/c` on the command line (MinGW transforms it to whatever is needed). I suspect it somehow doesn't parse ld/ar command line correctly and doesn't recognize the path as being a path or something like that.

What exactly do I need to download to get the version you're using? I have a dozen versions of different MinGWs here but so far I couldn't reproduce this behaviour in a couple of them I tried.

catalinr replied:
`git version` returns `git version 2.11.0.windows.1`. I guess it was taken from the regular git download site... try this: https://github.com/git-for-windows/git/releases/tag/v2.11.0.windows.1
It should come with its own bash which knows `pwd -W`.

The compiler is the one that came with Qt, see the instructions for the online installer. I think it was in a "Tools" section, but it should not be difficult to find among the other items since it was the latest (only?) gcc/mingw one. I used Qt 5.11 BTW.

vadz commented on include/wx/msw/apptrait.h:
```diff
@@ -87,6 +87,31 @@ class WXDLLIMPEXP_CORE wxGUIAppTraits : public wxGUIAppTraitsBase
     virtual bool WriteToStderr(const wxString& WXUNUSED(text)) { return false; }
 };
 
+#elif defined(__WXQT__)
```
This header starts getting really messy. I wonder if it could be better to have port-specific (rather than platform-specific) traits classes, at least for QT?

vadz commented on build/bakefiles/files.bkl:
```diff
@@ -360,7 +360,6 @@ IMPORTANT: please read docs/tech/tn0016.txt before modifying this file!
     src/qt/dcscreen.cpp
     src/qt/defs.cpp
     src/qt/dialog.cpp
-    src/unix/dialup.cpp
```
Also need to modify `build/files` (from which this one is normally generated by running `build/upmake` -- which should also update CMake files).

vadz reviewed (commented):
Some of the changes here (removing Unix files from QT files list; adding QT tests to the code defining `__WXMSW__`; DLL import fix for `wxTaskBarIcon`) are clearly fine and could be merged immediately if they could be extracted in separate commits. Other changes are less obviously correct though (notably event loop stuff) and I hesitate to apply them.

If you could refactor this PR to contain only the obviously correct changes, it would be nice as it could be merged immediately, leaving the other stuff for later.

catalinr committed:
removed Unix source files from QT

catalinr committed:
removed Qt value of TOOLKIT, kept only QT

catalinr committed:
more changes to accept mingw64

catalinr committed:
under Windows use win format for pwd and custom paths, required by ar and ld

catalinr committed:
removed leading spaces of a code line

catalinr committed:
DLL import fix for wxTaskBarIcon under QT

catalinr committed:
isolate Unix-only code with if guards

catalinr committed:
use the generic implementation of wxDoLaunchDefaultBrowser under wxQT

catalinr committed:
do not allow HAS_LOAD_FROM_RESOURCE for wxQT, at least for now

catalinr committed:
do not define __WXMSW__ and __WXQT__ at the same time

catalinr committed:
wxQT Windows build fix, rather incorrect

catalinr committed:
added wxQT declaration of wxGUIAppTraits

catalinr committed:
dll build fix for WXQT, DoEnableSystemTheme

removing 'inline' did fix:
warning: 'virtual void wxSystemThemedControlBase::DoEnableSystemTheme(bool, wxWindow*)' redeclared without dllimport attribute after being referenced with dll linkage

but introduced
multiple definition of `wxSystemThemedControlBase::DoEnableSystemTheme(bool, wxWindow*)'.

The proposed form gets no complaints

catalinr commented:
Bump.
The _needed work_ is mostly done. Every help will be appreciated for the remaining issues .

vadz commented:
Damn, I've totally forgotten about this one. It's probably going to conflict with #900, right?

Anyhow, I'll merge it if you think it's ok (i.e. it doesn't break anything that works currently), thanks.

vadz commented:
Sorry, I still didn't have time to actually check what's going on here, but I could at least resolve the (trivial) conflicts and commit your changes, hopefully more people could have a look at this now.

Thanks!
